File: Data Structures & Algorithms/count-paths/submission-0.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:

        cache = {}

        def unique(r,c):
            if (r,c) in cache:
                return cache[(r,c)]
            if r == m-1 and c == n-1:
                return 1
            if r == m or c == n:
                return 0
            cache[(r,c)] = unique(r+1,c) + unique(r,c+1)
            return cache[(r,c)]

        return unique(0,0)
        

File: Data Structures & Algorithms/coin-change/submission-0.js
class Solution {
    /**
     * @param {number[]} coins
     * @param {number} amount
     * @return {number}
     */
    coinChange(coins, amount) {
        const cache = {};
        function getTotalCoins(sum){
            if(sum in cache) return cache[sum];
            if(sum === amount) return 0;
            if(sum > amount)return Infinity;

            let res = Infinity;
            for (let coin of coins) {
                res = Math.min(res, 1 + getTotalCoins(sum + coin));
            }
            cache[sum] = res;
            return cache[sum];
        }

        const minCoins = getTotalCoins(0);
        return minCoins === Infinity ? -1 : minCoins;


    }
}
